### PR TITLE
PR: Improve cursor position history (Editor)

### PR DESCRIPTION
**Pull request by impact27**
Rework cursor position history

## Description of Changes

* [ ] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

Current cursor history loops through the entire history at each cursor position change. This PR avoids that.

### Issue(s) Resolved

Fixes #

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:

### CODE REVIEW

**Comment by ccordoba12** [comment]
Thanks a lot for all the improvements @impact27! I'm going to ask @steff456 to take a look at this one because she implemented this functionality.

**Comment by impact27** [comment]
I will update that to use cursors instead of positions - that way if a line above the cursor is modified, the right position will be retrieved

**Comment by ccordoba12** [comment]
@impact27, this PR has conflicts now.

**Comment by ccordoba12** [comment]
@impact27, you said in the description

> Current cursor history loops through the entire history at each cursor position change. This PR avoids that.

Could you expand on this? Does this rework helps to improve performance?

**Comment by impact27** [comment]
looping through the entire history at each cursor move degrades performance over time when the history is long

File: spyder/plugins/editor/plugin.py [hunk 1]
```diff
@@ -2633,46 +2642,92 @@ def go_to_last_edit_location(self):
                 if position < editor.document().characterCount():
                     editor.set_cursor_position(position)
 
+    def _pop_next_cursor_diff(self, history, current_filename, current_cursor):
+        """Get the next cursor from history that is different from current."""
+        while history:
```
Comment (steff456) [inline spyder/plugins/editor/plugin.py:2647]: Instead of doing a linear search on the list, you could create a dictionary and then just index it

**Review by steff456 — CHANGES_REQUESTED** [review]
I'm curious to know the time improvement taking into account these changes, because I'm seeing that you are creating more iterations over the list.

File: spyder/plugins/editor/plugin.py (hunk 1 above)
Comment (impact27) [reply spyder/plugins/editor/plugin.py:2647]: I am not sure what you mean. If you use a dictionary you loose the ordering, which is critical for an undo stack. On top of that this loop should mostly terminate after one iteration. Could you expand on your comment?

Comment (steff456) [reply spyder/plugins/editor/plugin.py:2647]: My bad, I didn't see that before. This looks great

**Review by steff456 — APPROVED** [review]
@ccordoba12 this is working as expected, I also tested it manually, I don't know if you have further comments. For me, this looks good to merge

**Comment by ccordoba12** [comment]
Thanks  @impact27 for your work on this and @steff456 for the review!

**Pull request merged**